### DISCUSSION TITLE
Support Kryoflux disk images

### DISCUSSION THREAD

**Discussion by Flyers80** [86Box/86Box#4257] [Ideas]
In other emulators there is already (Winuae / Steem / Caprice / Hatari)
it would be great if it were implemented in 86box too.
It is a single file that is more compact than those generated by supercard pro (* .scp) and can be used for copy-protected discs.

**Comment by OBattler** [comment]
Is there any kind of documentation of this format?

**Comment by Flyers80** [comment]
No :( , however the SPS decoder library contains the CT Raw format decoder and it's available as source code : http://www.softpres.org/_media/files:spsdeclib_5.1_windows.zip?id=download&cache=cache
And other documentation : http://www.softpres.org/download

**Comment by OBattler** [comment]
I think I looked at that once and it comes with full emulation of a WD FDC that is completely different from the NEC µPD765-compatible FDC's that PC and compatibles use.

**Comment by Flyers80** [comment]
ok, I deduce that it is not possible to apply it, we hope that at least support for the * .psi format is possible
thanks anyway :)

**Comment by OBattler** [comment]
Yes, .PSI support is currently in the plans.

**Comment by Kappa971** [comment]
Hi, I found some Kryoflux files for some copy-protected DOS games from [Myrient](https://myrient.erista.me/files/Total%20DOS%20Collection/Games/Images/Floppy/), specifically Bubble Bobble, Arkanoid, Arkanoid II, and Rambo III. The files are split into raw tracks, and I used the HxCFloppyEmulator program to convert them to the MFM file format.
Bubble Bobble and Arkanoid II work, but [Arkanoid](https://myrient.erista.me/files/Total%20DOS%20Collection/Games/Images/Floppy/1988/Arkanoid%20%281988%29%28Taito%20Software%20Inc.%29%20%5Bkfx%5D%5B360k%5D%5Bcp%5D%5B%21%5D.zip) and [Rambo III](https://myrient.erista.me/files/Total%20DOS%20Collection/Games/Images/Floppy/1989/Rambo%20III%20%281989%29%28Taito%20America%20Corporation%29%20%5Bkfx%5D%5B360k%5D%5Bcp%5D%5B%21%5D.zip) (disk 1 needs to be converted to 86f because it needs writing) don't detect the KEY Disk.
I found the same games already converted to MFM [here](https://archive.org/details/20230608_20230608) (the raw files are also available). I tried Arkanoid and it works (I haven't tried Rambo III yet).

I'd like to understand what's happening here because I can't find much information online. Why do some protected games work in 86Box when converted with HxCFloppyEmulator while others don't?

EDIT
I did the same conversion with HxCFloppyEmulator from raw files to mfm of Test Drive, Test Drive 2 (and its dlcs) and Lemmings and they work too (these games also have copy protection).

It would be nice if 86Box could convert the raw kryoflux format to .mfm and .86f without the need for additional programs. HxCFloppyEmulator doesn't seem to be reliable with all copy-protected images, or does something 86Box doesn't like, so if 86Box handled this properly, it would be less of a hassle for everyone to get working mfm images.

**Comment by Kappa971** [comment]
I managed to create a working Arkanoid floppy image from the Kryoflux .raw files.
I downloaded the [PCE emulator](http://www.hampa.ch/pub/pce/pre/) (never heard of it), which includes tools for converting floppy images. I used the command-line program `pfi.exe` to convert the image to .pri (PCE Raw Image) format.
86Box doesn't support this format, but I found a program ([DiskImageTool](https://github.com/Digitoxin1/DiskImageTool?tab=readme-ov-file)) to convert this image to 86f. Loading this 86f image, the game works.

The current problem is that if you encounter Kryoflux .raw or .scp files, there's currently no immediate way to convert and use these files in 86Box. A built-in tool to convert from .raw tracks or .scp to .mfm (if it works) or 86f would be useful.

**Reply by BlueRain-debug** [reply]
afaik GW supports converting between many different floppy disk image formats.

**Reply by Kappa971** [reply]
I can't get any formats other than .scp to work, but 86Box doesn't support the .scp format (nor does it allow you to convert it).

However, this isn't about helping me... rather, it would be very helpful to create a dedicated tool for 86Box to help everyone. 
Besides the HxC Floppy Emulator which failed in this case (with .mfm, 86Box doesn't support .hfe), there are no other comprehensive programs that allow you to convert raw tracks to a format usable by 86Box.

**Reply by OBattler** [reply]
Are you using my modified version of HxC Floppy Emulator that produces modified MFM flies that don't round track bitcell counts to the nearest byte and that store RPM and data rate per track instead of per disk? Otherwise, I can see why it doesn't work.

**Reply by Kappa971** [reply]
No... where can I get your version?

**Reply by Kappa971** [reply]
I can't find your version. Why don't you fork HxC Floppy Emulator? You could remove all the stuff 86Box doesn't need, fix the .mfm format and/or add the 86f format for the conversion, or create a new command-line program like pfi.exe to convert from raw tracks to 86f. I think it would be an essential tool for video game emulation purposes.

**Reply by OBattler** [reply]
Here you go: https://github.com/user-attachments/files/25993824/hxc_with_modded_mfm.zip .

> Why don't you fork HxC Floppy Emulator?

Because I already have enough work from 86Box alone.

**Reply by OBattler** [reply]
Also, for what it's worth, the author of HxC started adding 86F support but it's still preliminary and buggy.

**Reply by Kappa971** [reply]
> Here you go: https://github.com/user-attachments/files/25993824/hxc_with_modded_mfm.zip .

Thanks, I will try it.

> Because I already have enough work from 86Box alone.

Even a small program like those included with PCE would be useful for converting, in this case, to .86f (forget about the HxC fork, which I understand could be too time-consuming).
I learned about this 86Box format thanks to some copy-protected games. From the information I've found online, it seems to be potentially the most advanced for "special" floppy disk images, but the paradox is that there's currently no tool to convert Kryoflux raw images to this format, only from .mfm, .td0, or other formats that have lost the information .86f can contain (if I understand correctly), although this might only be an issue with some software/games with weird copy protections.

I'm just asking you to provide us the tools to use what you've created 😅.

> Also, for what it's worth, the author of HxC started adding 86F support but it's still preliminary and buggy.

The latest beta version seems to only read 86f but it gives some weird information so yeah, it's a bit buggy 😅. It's probably not a priority for him.

In my opinion, it's not a good idea to rely on or wait for a "third-party" program focused on other things to add support for your image format.
I understand that 86Box requires a lot of work, but this is part of 86Box in a way.

**Comment by OBattler** [comment]
[hxc_with_modded_mfm.zip](https://github.com/user-attachments/files/25993824/hxc_with_modded_mfm.zip)
Here's my modified HxC. I actually proposed to the author to adopt my modified MFM format but he said no.

**Reply by Kappa971** [reply]
I converted the image to .mfm using your modified version. The game now works, but (a small note you probably already know) the image is only readable by your program and 86Box.

I found this: [Arkanoid (1988) (360K) [cp] [!].zip](https://github.com/user-attachments/files/25996180/Arkanoid.1988.360K.cp.zip)
It's an .mfm image of the same game converted by someone and uploaded to Archive.org. I can't figure out why this works... He couldn't have used HxC Floppy Emulator (otherwise the image wouldn't work), nor your modified version (this file is recognized by other programs like DiskImageTool, so it can't be from your version), and I haven't found any other programs besides DiskImageTool that convert images to .mfm (this also breaks the game's protection), so how did he do it?

In the raw track format ([Arkanoid[Flux].zip](https://github.com/user-attachments/files/25996190/Arkanoid.Flux.zip)) there are two files: a log file and a strange text file called `tracklayout.txt` with inside
```
0-15:80,50,80
16-16:80,50,[10,70,70,70,70,70,70,70,70].80,50,80
17-39:80,50,80
```
Did he perhaps use this file with HxC Floppy Emulator? What do those ​​80, 50, and 70 values mean?

**Reply by OBattler** [reply]
Those values could be bitcell position offsets.

**Reply by Kappa971** [reply]
Test Drive floppy image crashes your version of HxC (not the standard one):
[Test Drive v1.1 (1987)(Accolade, Inc.) [kfx][360k][cp][!].zip](https://github.com/user-attachments/files/26005782/Test.Drive.v1.1.1987.Accolade.Inc.kfx.360k.cp.zip)

Not a big problem, copy protection works even when converting the image with the standard version.

**Reply by Kappa971** [reply]
This old version of HxC is particularly unstable... In addition to the fact that some images like the one above crash the program, it also becomes unstable and crashes after converting 4-5 images.

Another note, when converting from modified .mfm to .86f via 86Box, another track is added to the image (is it normal?).
